lang/cs/Org.Apache.REEF.Network/Elastic/Topology/Logical/Impl/FlatTopology.cs

```diff
@@ -0,0 +1,448 @@
+                _nodesWaitingToJoinTopologyNextIteration = new HashSet<string>();
+            }
+        }
+
+        /// <summary>
+        /// Reconfigure the topology in response to some event.
+        /// </summary>
+        /// <param name="taskId">The task id responsible for the topology change</param>
+        /// <param name="info">Some additional topology-specific information</param>
+        /// <param name="iteration">The optional iteration number in which the event occurred</param>
+        /// <returns>One or more messages for reconfiguring the Tasks</returns>
+        public IList<IElasticDriverMessage> Reconfigure(
+            string taskId,
+            Optional<string> info,
+            Optional<int> iteration)
+        {
+            if (taskId == _rootTaskId)
+            {
+                throw new NotImplementedException("Failure on master not supported yet.");
+            }
+
+            List<IElasticDriverMessage> messages = new List<IElasticDriverMessage>();
+
+            lock (_lock)
+            {
+                int iter;
+
+                if (info.IsPresent())
+                {
+                    iter = int.Parse(info.Value.Split(':')[0]);
+                }
+                else
+                {
+                    iter = iteration.Value;
+                }
+
+                var children = _lostNodesToBeRemoved.ToList();
+                var update = new List<TopologyUpdate>()
+                {
+                    new TopologyUpdate(_rootTaskId, children)
+                };
+                var data = new FailureMessagePayload(update, StageName, OperatorId, -1);
+                var returnMessage = new ElasticDriverMessageImpl(_rootTaskId, data);
+
+                Log.Log(Level.Info, "Task {0} is removed from topology", taskId);
+                messages.Add(returnMessage);
+                _lostNodesToBeRemoved.Clear();
+            }
+
+            return messages;
+        }
+
+        /// <summary>
+        /// Log the final statistics of the operator.
+        /// This is called when the pipeline execution is completed.
+        /// </summary>
+        public string LogFinalStatistics()
+        {
+            return string.Format(
+                "\nAverage number of nodes in the topology of Operator {0}: {1}",
+                OperatorId,
+                _iteration >= 2 ? (float)_totNumberofNodes / (_iteration - 1) : _availableDataPoints);
+        }
+
+        private void BuildTopology()
+        {
+            IEnumerator<DataNode> iter =
+                _sorted ?
+                    _nodes.OrderBy(kv => kv.Key).Select(kv => kv.Value).GetEnumerator() :
+                    _nodes.Values.GetEnumerator();
+            var root = _nodes[_rootId];
+
+            while (iter.MoveNext())
+            {
+                if (iter.Current.TaskId != _rootId)
+                {
+                    root.AddChild(iter.Current);
+                }
+            }
+        }
```

Comment:
the entire method can be written in one line:
```C#
_nodes[_rootId].AddChildren(_nodes.Values.Where(node => node.TaskId != _rootId));
```
if we implement `DataNode.AddChildren(IEnumerable<DataNode> children)` and make `_nodes` a `SortedDictionary` (or an `IDictionary` that is either sorted or not depending on `_sorted` flag)

Reply:
Yep. Fixed.